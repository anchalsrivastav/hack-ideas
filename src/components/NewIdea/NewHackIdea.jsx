import React, { useRef, useState } from 'react'
import { Button, Form, Input, Select } from "antd";
import { useLocation } from 'react-router';
import { useEffect } from 'react/cjs/react.development';
import { NavLink } from 'react-router-dom';
import {formButton, formWrapper} from "./NewHackIdea.styles";
import {tags} from "../../constants";

const {Option} = Select;

const NewHackIdea = () => {
    const location = useLocation();
    const [hackListItems, setHackListItems] = useState([]);
    const formRef = useRef("");
    const errorMessages = [ {required: true, message: 'Please enter valid input'} ]
    const children = Object.entries(tags).map(([tagid, tagname], idx) => <Option key={tagid}>{tagname}</Option>);

    useEffect(() => {
        const hackList = location.state?.hackList ? location.state?.hackList : [];
        localStorage.setItem("hackIdea", JSON.stringify(hackList));
        setHackListItems([...hackList]);
    }, [location.state?.hackList]);

    const onSubmit = (values) => {
        const newHackIdeaList = [...hackListItems, {
            id: hackListItems.length + 1,
            creationDate: new Date(),
            title: values["title"],
            description: values["description"],
            tags: values["tags"],
            upvotes: 0,
        }]
        setHackListItems(newHackIdeaList);
        formRef.current?.resetFields();
        localStorage.setItem("hackIdea", JSON.stringify(newHackIdeaList));
    };

    return (
        <div style={formWrapper}>
            <Form ref={formRef} onFinish={onSubmit}>
                <h1>Add New Hack Item</h1>
                <Form.Item name="title" label="Title" rules={errorMessages}>
                    <Input />
                </Form.Item>
                <Form.Item name="description" label="Description" rules={errorMessages}>
                    <Input />
                </Form.Item>
                <Form.Item name="tags" label="Tags" rules={errorMessages}>
                    <Select
                        mode="tags"
                        placeholder="Please select Tags"
                        value={[]}
                        style={tags}
                    >
                        {children}
                    </Select>
                </Form.Item>
                <div>
                    <Button style={formButton} type="primary" htmlType="submit">Add HackIDEA</Button>
                    <Button style={formButton} type="default" htmlType="reset">Reset</Button>
                    <Button style={formButton}><NavLink to={'/dashboard'}>Go To Dashboard</NavLink></Button>
                </div>
            </Form>
        </div>
    )
}

export default NewHackIdea
