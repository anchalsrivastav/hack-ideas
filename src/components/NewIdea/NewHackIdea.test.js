import NewHackIdea from "./NewHackIdea";
import {render} from '@testing-library/react'
import {createMemoryHistory} from 'history'
import React from 'react';
import {Router} from 'react-router-dom';
import '@testing-library/jest-dom';

describe('New Hack Idea', () => {
    beforeEach(() => {
        Object.defineProperty(window, 'matchMedia', {
            writable: true,
            value: jest.fn().mockImplementation(query => ({
                matches: false,
                media: query,
                onchange: null,
                addListener: jest.fn(),
                removeListener: jest.fn(),
                addEventListener: jest.fn(),
                removeEventListener: jest.fn(),
                dispatchEvent: jest.fn(),
            })),
        });
    });

    it('renders new hack idea creation page', () => {
        const history = createMemoryHistory()
        const route = '/add';
        history.push(route)
        const {container, getByText} = render(
            <Router history={history}>
                <NewHackIdea />
            </Router>,
        )
        expect(getByText('Add New Hack Item')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    })
})
