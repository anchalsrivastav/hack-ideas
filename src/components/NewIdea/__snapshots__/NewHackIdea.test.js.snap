// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Hack Idea renders new hack idea creation page 1`] = `
<div>
  <div
    style="height: 100%; width: 100%; display: flex; justify-content: center; margin: 4rem;"
  >
    <form
      class="ant-form ant-form-horizontal"
    >
      <h1>
        Add New Hack Item
      </h1>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class="ant-form-item-required"
            for="title"
            title="Title"
          >
            Title
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <input
                class="ant-input"
                id="title"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class="ant-form-item-required"
            for="description"
            title="Description"
          >
            Description
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <input
                class="ant-input"
                id="description"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class="ant-form-item-required"
            for="tags"
            title="Tags"
          >
            Tags
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-select ant-select-multiple ant-select-show-search"
              >
                <div
                  class="ant-select-selector"
                >
                  <div
                    class="ant-select-selection-overflow"
                  >
                    <div
                      class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
                      style="opacity: 1;"
                    >
                      <div
                        class="ant-select-selection-search"
                        style="width: 0px;"
                      >
                        <input
                          aria-activedescendant="tags_list_0"
                          aria-autocomplete="list"
                          aria-controls="tags_list"
                          aria-haspopup="listbox"
                          aria-owns="tags_list"
                          autocomplete="off"
                          class="ant-select-selection-search-input"
                          id="tags"
                          role="combobox"
                          type="search"
                          value=""
                        />
                        <span
                          aria-hidden="true"
                          class="ant-select-selection-search-mirror"
                        >
                          
                          Â 
                        </span>
                      </div>
                    </div>
                  </div>
                  <span
                    class="ant-select-selection-placeholder"
                  >
                    Please select Tags
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <button
          class="ant-btn ant-btn-primary"
          style="margin: 1rem;"
          type="submit"
        >
          <span>
            Add HackIDEA
          </span>
        </button>
        <button
          class="ant-btn"
          style="margin: 1rem;"
          type="reset"
        >
          <span>
            Reset
          </span>
        </button>
        <button
          class="ant-btn"
          style="margin: 1rem;"
          type="button"
        >
          <a
            href="/dashboard"
          >
            Go To Dashboard
          </a>
        </button>
      </div>
    </form>
  </div>
</div>
`;
